sd/etcd/registrar.go

```diff
@@ -19,9 +25,27 @@ type Registrar struct {
 type Service struct {
 	Key           string // unique key, e.g. "/service/foobar/1.2.3.4:8080"
 	Value         string // returned to subscribers, e.g. "http://1.2.3.4:8080"
+	TTL           *TTLOption
 	DeleteOptions *etcd.DeleteOptions
 }
 
+// TTLOption allow setting a key with a TTL, and regularly refreshes the lease with a goroutine
+type TTLOption struct {
+	Heartbeat time.Duration
+	TTL       time.Duration
+}
+
+// NewTTLOption returns a TTLOption
+func NewTTLOption(heartbeat, ttl time.Duration) *TTLOption {
+	if heartbeat <= MinHeartBeatTime {
+		heartbeat = MinHeartBeatTime
+	}
+	if ttl <= heartbeat {
+		ttl = heartbeat * 3
+	}
+	return &TTLOption{heartbeat, ttl}
```

Comment:
Given that both heartbeat and ttl have the same type, it's easy to flip them around. It's a good idea to use struct literal initialization, e.g.

```go
return &TTLOption{
    heartbeat: heartbeat,
    ttl:       ttl,
}
```